# Laboratorium 5 — Pandas — Series, DataFrame, eksploracja danych
**Autor:** Marta Machacka

## Ćwiczenie 1: Series — tworzenie i operacje

In [8]:
import pandas as pd
import numpy as np

# Series z listy — domyślny indeks (0, 1, 2, ...)
oceny = pd.Series([4.5, 3.0, 5.0, 3.5, 4.0])
print(oceny.to_string())

0    4.5
1    3.0
2    5.0
3    3.5
4    4.0


### Zadanie 1a: Utwórz Series pensje z dict:

In [54]:
# Klucze: 'Anna', 'Jan', 'Ewa', 'Marek', 'Kasia'
# Wartości: 5500, 7200, 4800, 9100, 6300
pensje = pd.Series({'Anna' : 5500,
                    'Jan' : 7200,
                    'Ewa': 4800,
                    'Marek': 9100,
                    'Kasia': 6300})

### Zadanie 1b: Odpowiedz na pytania o pensje:

In [55]:
# 1. Wyświetl imię osoby zarabiającej najwięcej (użyj idxmax())
osoba_zarobek_max = pensje.idxmax()
print(f"Osoba zarabiająca najwięcej: {osoba_zarobek_max}\n")

# 2. Oblicz średnią pensję
srednia_pensja = pensje.mean()
print(f"Średnia pensja: {srednia_pensja} zł\n")

# 3. Wyświetl imiona osób zarabiających powyżej średniej (filtrowanie boolean)
maska = pensje > srednia_pensja
print(f"Osoby zarabiające powyżej średniej: {', '.join(pensje[maska].index)}\n")

# 4. Policz ile osób zarabia poniżej 6000 (użyj len() lub .sum())
maska = pensje.values < 6000
print(f"Osoby zarabiające poniżej 6000 zł: {len(pensje[maska])}")

Osoba zarabiająca najwięcej: Marek

Średnia pensja: 6580.0 zł

Osoby zarabiające powyżej średniej: Jan, Marek

Osoby zarabiające poniżej 6000 zł: 2


### Zadanie 1c: Oblicz pensje po zmianach:

In [56]:
# 1. Oblicz pensje po podwyżce 10% dla wszystkich (pensje * 1.10)
podwyzka_10 = pensje * 1.1

print(f"--- Pensje po podwyżce 10% ---\n{podwyzka_10.to_string()}")
# 2. Oblicz pensje po premii 500 zł dla wszystkich (pensje + 500)
pensje_po_premii = pensje + 500
print(f"\n--- Pensje po premii 500 zł ---\n{pensje_po_premii.to_string()}")

# 3. Oblicz nową pensję = (pensja * 1.10) + 500
nowe_pensje = (pensje * 1.10) + 500
print(f"\n--- Nowe pensje (+ 10% + 500 zł) ---\n{nowe_pensje.to_string()}")
# 4. Wyświetl ranking nowych pensji od najwyższej (użyj sort_values)

print(f"--- Ranking  --- \n{nowe_pensje.sort_values().to_string()}")

--- Pensje po podwyżce 10% ---
Anna      6050.0
Jan       7920.0
Ewa       5280.0
Marek    10010.0
Kasia     6930.0

--- Pensje po premii 500 zł ---
Anna     6000
Jan      7700
Ewa      5300
Marek    9600
Kasia    6800

--- Nowe pensje (+ 10% + 500 zł) ---
Anna      6550.0
Jan       8420.0
Ewa       5780.0
Marek    10510.0
Kasia     7430.0
--- Ranking  --- 
Ewa       5780.0
Anna      6550.0
Kasia     7430.0
Jan       8420.0
Marek    10510.0


## Ćwiczenie 2: DataFrame — tworzenie i atrybuty

Krok 1 — DataFrame z dict

In [57]:
# Dane o produktach sklepu internetowego
sklep = pd.DataFrame({
'produkt': ['Laptop', 'Tablet', 'Smartfon', 'Słuchawki', 'Monitor'],
'cena': [3500, 1800, 2500, 350, 2200],
'stan_magazynowy': [45, 120, 200, 500, 75],
'kategoria': ['Komputery', 'Mobilne', 'Mobilne', 'Akcesoria', 'Komputery']
})

### Zadanie 2a: Zbadaj DataFrame sklep

In [58]:
# 1. Wyświetl kształt tabeli (shape)
sklep.shape

(5, 4)

In [61]:
# 2. Wyświetl nazwy kolumn (columns)
sklep.columns

Index(['produkt', 'cena', 'stan_magazynowy', 'kategoria'], dtype='str')

In [62]:
# 3. Wyświetl typy danych w każdej kolumnie (dtypes)
sklep.dtypes

produkt              str
cena               int64
stan_magazynowy    int64
kategoria            str
dtype: object

In [63]:
# 4. Wyświetl pierwsze 3 wiersze (head)
sklep.head(3)

,produkt,cena,stan_magazynowy,kategoria
0,Laptop,3500,45,Komputery
1,Tablet,1800,120,Mobilne
2,Smartfon,2500,200,Mobilne


### Zadanie 2b: Dodaj nowe kolumny do DataFrame sklep

In [64]:
# 1. Dodaj kolumnę 'wartosc_magazynu' = cena * stan_magazynowy
sklep['wartosc_magazynu'] = (sklep['cena'] * sklep['stan_magazynowy']).round(1)


In [65]:
# 2. Dodaj kolumnę 'cena_z_vat' = cena * 1.23
sklep['cena_z_vat'] = (sklep['cena'] * 1.23).round(1)

In [66]:
# 3. Wyświetl cały DataFrame z nowymi kolumnami
print(sklep)

     produkt  cena  stan_magazynowy  kategoria  wartosc_magazynu  cena_z_vat
0     Laptop  3500               45  Komputery            157500      4305.0
1     Tablet  1800              120    Mobilne            216000      2214.0
2   Smartfon  2500              200    Mobilne            500000      3075.0
3  Słuchawki   350              500  Akcesoria            175000       430.5
4    Monitor  2200               75  Komputery            165000      2706.0


Krok 2 — Wczytanie CSV

In [67]:
# Wczytaj dataset tips
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv'
tips = pd.read_csv(url)

### Zadanie 2c: Podstawowa eksploracja datasetu tips

In [68]:
# 1. Wyświetl ile wierszy i kolumn ma dataset (shape)
tips.shape

(244, 7)

In [69]:
# 2. Wyświetl ostatnie 5 wierszy (tail)
tips.tail(5)

,total_bill,tip,sex,smoker,day,time,size
239,29.03,5.92,Male,No,Sat,Dinner,3
240,27.18,2.00,Female,Yes,Sat,Dinner,2
241,22.67,2.00,Male,Yes,Sat,Dinner,2
242,17.82,1.75,Male,No,Sat,Dinner,2
243,18.78,3.00,Female,No,Thur,Dinner,2


In [70]:
# 3. Wyświetl 3 losowe wiersze (sample)
tips.sample(3)

,total_bill,tip,sex,smoker,day,time,size
220,12.16,2.2,Male,Yes,Fri,Lunch,2
187,30.46,2.0,Male,Yes,Sun,Dinner,5
177,14.48,2.0,Male,Yes,Sun,Dinner,2


In [71]:
# 4. Wyświetl typy danych w każdej kolumnie (dtypes)
tips.dtypes

total_bill    float64
tip           float64
sex               str
smoker            str
day               str
time              str
size            int64
dtype: object

## Ćwiczenie 3: Eksploracja datasetu — samodzielna analiza

In [72]:
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/penguins.csv'
penguins = pd.read_csv(url)

### Zadanie 3a: Pierwszy rzut oka

In [73]:
# 1. Wyświetl head(10)
penguins.head(10)

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
0,Adelie,Torgersen,39.1,18.7,181.0,3750.0,MALE
1,Adelie,Torgersen,39.5,17.4,186.0,3800.0,FEMALE
2,Adelie,Torgersen,40.3,18.0,195.0,3250.0,FEMALE
3,Adelie,Torgersen,NaN,NaN,NaN,NaN,NaN
4,Adelie,Torgersen,36.7,19.3,193.0,3450.0,FEMALE
5,Adelie,Torgersen,39.3,20.6,190.0,3650.0,MALE
6,Adelie,Torgersen,38.9,17.8,181.0,3625.0,FEMALE
7,Adelie,Torgersen,39.2,19.6,195.0,4675.0,MALE
8,Adelie,Torgersen,34.1,18.1,193.0,3475.0,NaN
9,Adelie,Torgersen,42.0,20.2,190.0,4250.0,NaN


In [74]:
# 2. Wyświetl shape
penguins.shape

(344, 7)

In [76]:
# 3. Uruchom info() — ile brakujących wartości widzisz?
penguins.info()

print(f"\n -- Braki w każdej kolumnie -- ")
print(penguins.isnull().sum().to_string())

print(f"\n -- Łączna ilość brakujących wartości {penguins.isnull().sum().sum()} -- ")

<class 'pandas.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    str    
 1   island             344 non-null    str    
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    str    
dtypes: float64(4), str(3)
memory usage: 18.9 KB

 -- Braki w każdej kolumnie -- 
species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11

 -- Łączna ilość brakujących wartości 19 -- 


### Zadanie 3b: Statystyki opisowe

In [77]:
# 1. Uruchom describe() — co mówią ci średnie i odchylenia?
penguins.describe()


,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g
count,342.000000,342.000000,342.000000,342.000000
mean,43.921930,17.151170,200.915205,4201.754386
std,5.459584,1.974793,14.061714,801.954536
min,32.100000,13.100000,172.000000,2700.000000
25%,39.225000,15.600000,190.000000,3550.000000
50%,44.450000,17.300000,197.000000,4050.000000
75%,48.500000,18.700000,213.000000,4750.000000
max,59.600000,21.500000,231.000000,6300.000000


Średnie: 
- bill_length_mm ≈ 43.9 mm
- bill_depth_mm ≈ 17.15 mm
- flipper_length_mm ≈ 200.9 mm
- body_mass_g ≈ 4202 g
Dają obraz wymiarów "przeciętnego" osobnika - średni dziób ~44 mm, płetwa ~201 mm, masa ~4.2 kg.

Odchylenia:
- bill_length_mm: ~5.46 → umiarkowana zmienność
- bill_depth_mm: ~1.97 → mała zmienność (dość stabilna cecha)
- flipper_length_mm: ~14.06 → umiarkowana zmienność
- body_mass_g: ~802 → duża zmienność
Największa zmienność jest w wadze osobników


In [78]:
# 2. Uruchom describe(include='all') — co dodają kolumny tekstowe?
penguins.describe(include='all')

,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex
count,344,344,342.000000,342.000000,342.000000,342.000000,333
unique,3,3,NaN,NaN,NaN,NaN,2
top,Adelie,Biscoe,NaN,NaN,NaN,NaN,MALE
freq,152,168,NaN,NaN,NaN,NaN,168
mean,NaN,NaN,43.921930,17.151170,200.915205,4201.754386,NaN
std,NaN,NaN,5.459584,1.974793,14.061714,801.954536,NaN
min,NaN,NaN,32.100000,13.100000,172.000000,2700.000000,NaN
25%,NaN,NaN,39.225000,15.600000,190.000000,3550.000000,NaN
50%,NaN,NaN,44.450000,17.300000,197.000000,4050.000000,NaN
75%,NaN,NaN,48.500000,18.700000,213.000000,4750.000000,NaN


Dla kolumn tekstowych describe(include='all') pokazuje strukturę kategorii, a nie wartości liczbowe. 
Widoczne są: liczba niepustych wartości, liczbę unikalnych wartości, najczęściej występującą wartość (dominantę/modę), ile razy występuje dana wartość.

SPECIES:
- count = 344 → brak braków
- unique = 3 → są 3 gatunki
- top = Adelie → najczęstszy gatunek
- freq = 152 → występuje 152 razy

ISLAND:

- unique = 3
- top = Biscoe → najczęstsza wyspa

SEX:

- count = 333 → są braki
- unique = 2 → MALE / FEMALE
- top = MALE
- freq = 168

In [79]:
# 3. Jaka jest mediana masy ciała (body_mass_g)?
print(f"Mediana masy ciała: {penguins['body_mass_g'].median()} g")

Mediana masy ciała: 4050.0 g


## Zadanie 3c: Rozkłady wartości

In [80]:
# 1. Ile jest pingwinów każdego gatunku? (value_counts na 'species')
print(penguins['species'].value_counts().to_string())

species
Adelie       152
Gentoo       124
Chinstrap     68


In [81]:
# 2. Ile pingwinów na każdej wyspie? (value_counts na 'island')
print(penguins['island'].value_counts().to_string())

island
Biscoe       168
Dream        124
Torgersen     52


In [82]:
# 3. Rozkład płci (value_counts na 'sex') — czy jest zbalansowany?
print(penguins['sex'].value_counts().to_string())

sex
MALE      168
FEMALE    165


Rozkład płci jest zbalansowany - jest tylko o 3 samców więcej niż samic

In [83]:
# 4. Jaki procent to gatunek Adelie? (normalize=True)
print(f"Gatunek Adelie stanowi {(penguins['species'].value_counts(normalize=True)['Adelie']).round(4)*100:.2f}% wszystkich pingwinów.")

Gatunek Adelie stanowi 44.19% wszystkich pingwinów.


## Zadanie 3d: Brakujące dane

In [84]:
# 1. Policz brakujące wartości w każdej kolumnie (isna().sum())
print(f"\n -- Braki w każdej kolumnie -- ")
print(penguins.isna().sum().to_string())


 -- Braki w każdej kolumnie -- 
species               0
island                0
bill_length_mm        2
bill_depth_mm         2
flipper_length_mm     2
body_mass_g           2
sex                  11


In [85]:
# 2. Jaki procent danych brakuje w kolumnie 'sex'?
# Podpowiedź: braki / len(penguins) * 100
print(f"W kolumnie 'sex' brakuje {(penguins['sex'].isna().sum()/len(penguins) * 100).round(2)}% danych.")

W kolumnie 'sex' brakuje 3.2% danych.


## Zadanie 3e: Wnioski (refleksja)

- Dataset zawiera 344 pingwinów z 3 gatunków:
Adelie    →   152,
Gentoo    →   124,
Chinstrap    →   68

- Najczęstszy gatunek to Adelie (152 osobników, 44.19%)

- Brakuje danych w kolumnach: species, island, bill_length_mm,
bill_depth_mm, flipper_length_mm, body_mass_g, sex 
(łącznie 19 braków)

- Średnia masa ciała to około 4201.75 g (mediana: 4050.0 g)

- Pingwiny mieszkają na 3 wyspach, najliczniej na Biscoe

## Ćwiczenie 4: Pytania biznesowe

In [86]:
tips = pd.read_csv('https://raw.githubusercontent.com/mwaskom/seaborn-data/master/tips.csv')
# Dodaj kolumnę z procentem napiwku
tips['tip_pct'] = (tips['tip'] / tips['total_bill'] * 100).round(1)


### Zadanie 4a

In [87]:
# 1. Oblicz średni rachunek (total_bill)
print(f"Średni rachunek: {tips['total_bill'].mean():.2f}")

Średni rachunek: 19.79


In [88]:
# 2. Oblicz średni napiwek w procentach (tip_pct)
print(f"Średni napiwek: {tips['tip_pct'].mean():.2f}%")

Średni napiwek: 16.08%


In [89]:
# 3. Wyświetl który dzień ma najwięcej rachunków (value_counts na 'day')
print(f"Dzień z największą ilością rachunków: {tips['day'].value_counts().idxmax()}")


Dzień z największą ilością rachunków: Sat


In [90]:
# 4. Policz ile rachunków jest od palaczy (value_counts na 'smoker')
print(f"Ilość rachunków od palaczy: {tips['smoker'].value_counts()['Yes']}")



Ilość rachunków od palaczy: 93


In [91]:
# 5. Wyświetl describe() tylko dla kolumny 'tip' — jaka jest mediana napiwku?
print(tips['tip'].describe().to_string())

count    244.000000
mean       2.998279
std        1.383638
min        1.000000
25%        2.000000
50%        2.900000
75%        3.562500
max       10.000000


Medniana napiwku wynosi 2.9

In [92]:
# 6. Jaki procent rachunków pochodzi z obiadu (Dinner)?
# Podpowiedź: value_counts(normalize=True) na kolumnie 'time'
print(f"Rachunki z obiadu stanowią {((tips['time'].value_counts(normalize=True)['Dinner'])*100):.2f}% wszystkich rachunków.")


Rachunki z obiadu stanowią 72.13% wszystkich rachunków.


In [93]:
# 7. Ile unikalnych rozmiarów grup jest w danych? (nunique na 'size')
print(f"Jest {tips['size'].nunique()} unikalnych rozmiarów grup.")

Jest 6 unikalnych rozmiarów grup.
